### PR TITLE
added oauth2 functionality behind feature switch

### PR DESCRIPTION
**Pull request by whitfield-mj**
fixed auth callback use to get. Throw error if required role is not present

Fixed tests

### CODE REVIEW

File: server/authentication/auth.js [hunk 1]
```diff
@@ -27,22 +30,22 @@ passport.deserializeUser(function(user, done) {
...
-            if (isEmpty(user)) {
-                return done(null, false, {message: 'Incorrect username or password'});
-            }
-
-            return done(null, user);
-        } catch (error) {
-            return done(null, false, {message: 'A system error occured; please try again later'});
-        }
-    });
+function init(signInService, userService, audit, authStrategy) {
```
Comment (petergphillips) [inline server/authentication/auth.js]: don't need to pass in `config.authStrategy` here as `auth` already has access to `config`

File: server/authentication/strategies.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
+const fiveMinutesBefore = require('../utils/fiveMinutesBefore');
+const {getIn} = require('../utils/functionalHelpers');
+const allowedRoles = require('./roles');
+const logger = require('../../log');
+
+module.exports = function(signInService, userService, audit) {
```
Comment (petergphillips) [inline server/authentication/strategies.js]: I wouldn't change it now, but this looks to be over engineered.  Factoring out the strategies into their own file would be good if the feature switch is designed to be long lived and we be using both for a long period of time.  In our case it makes it harder to take the code out when the feature switch is removed.

File: server/services/userService.js [hunk 3]
```diff
@@ -1,6 +1,28 @@
...
+    async function getUserProfile(token, refreshToken, username) {
+        logger.info(`User profile for: ${username}`);
+        const nomisClient = nomisClientBuilder(token);
+
+        const [profile, roles] = await Promise.all([
+            nomisClient.getLoggedInUserInfo(),
+            getAllRoles({token})
+        ]);
+
+        logger.info(`User profile success - token: ${token}`);
```
Comment (petergphillips) [inline server/services/userService.js]: don't think it is a good idea to log the user's `token` here

File: server/authentication/strategies.js (hunk 2 above)
Comment (whitfield-mj) [reply server/authentication/strategies.js]: Discussed and renamed file

File: server/services/userService.js (hunk 3 above)
Comment (whitfield-mj) [reply server/services/userService.js]: Agreed

**Pull request merged**